### PR TITLE
Try to fix GROUP BY _shard_num in a different way

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Interpreters/ClusterProxy/SelectStreamFactory.cpp [hunk 1]
```diff
@@ -73,6 +75,89 @@ SelectStreamFactory::SelectStreamFactory(
 namespace
 {
 
+/// Special support for the case when `_shard_num` column is used in GROUP BY key expression.
+/// This column is constant is a constant for shard.
```
Comment (amosbird) [inline src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:79]: typo

**Review by azat — COMMENTED** [review]
@KochetovNicolai thanks, looks good!

By the way, right now the are still few things that does not work:
- `SELECT _shard_num s, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY s` - https://github.com/ClickHouse/ClickHouse/pull/16947/commits/e450ac2f0c8a4296d95f19430bb04369133b405a
- `select materialize(_shard_num), * from remote('127.{1,2}', system.one) limit 1 by dummy` - https://github.com/ClickHouse/ClickHouse/pull/16947/commits/51f8628005625997dcc624892b010d911e7d33bf

I left my patches for these examples, but maybe you will come up with a better fix.

**Comment by KochetovNicolai** [comment]
@azat I think I have fixed your cases. Do you have more?

**Comment by azat** [comment]
>I think I have fixed your cases. Do you have more?

No, seems that it covers all the cases (at least those that I know), thanks!

**Pull request merged**